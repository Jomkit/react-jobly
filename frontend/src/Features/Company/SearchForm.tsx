import React, { useState } from 'react';
import JoblyApi from '../../api.ts';
import { Col, FormGroup, Input, Label, Row } from 'reactstrap'
import { companyFormInterface } from '../../types.ts';
const SearchForm = ({ setData }: { setData: Function }) => {
    const initialState: companyFormInterface = {
        name: "", 
        minEmployees: 0,
        maxEmployees: 10000
    }
    const [ formData, setFormData ] = useState(initialState);

    // Need to apply filter to form data
    async function getFiltered(filterParams: companyFormInterface) {
        try{
            const res = await JoblyApi.getCompanies(filterParams);
            setData(res);
        } catch(e: any){
            console.log("Oops, something went wrong...");
            console.error(e);
        }
    }
    
    function handleChange(evt: React.ChangeEvent<HTMLInputElement>){
        const { name, value } = evt.target;
        setFormData((data: companyFormInterface) => ({
            ...data,
            [name]: value,
        }));
    }

    function handleSubmit(evt: React.FormEvent){
        evt.preventDefault();
        
        getFiltered(formData);
    }

  return (
    <form className='bg-white mx-5 border border-3 border-black rounded p-1' onSubmit={handleSubmit}>
        <Row className='m-lg-4 align-items-center'>
            <Col lg={6}>
                <FormGroup>
                    <Label
                        className='visually-hidden'
                        for='name'
                    >
                        Name
                    </Label>
                    <Input
                    id='name'
                    name='name'
                    placeholder='search companies'
                    type="text"
                    value={formData.name}
                    onChange={handleChange}
                    />            
                </FormGroup>
            </Col>
            <Col lg={2}>
                <FormGroup floating>
                    <Input 
                        id="minEmployees"
                        name="minEmployees"
                        placeholder='0'
                        min={0}
                        type="number"
                        bsSize='sm'
                        value={formData.minEmployees}
                        onChange={handleChange}
                    />
                    <Label for="minEmployees">Min Employees</Label>
                </FormGroup>
            </Col>
            <Col lg={2}>
                <FormGroup floating>
                    <Input 
                        id="maxEmployees"
                        name="maxEmployees"
                        placeholder='0'
                        min={0}
                        type="number"
                        bsSize='sm'
                        value={formData.maxEmployees}
                        onChange={handleChange}
                    />
                    <Label for="maxEmployees">Max Employees</Label>
                </FormGroup>
            </Col>

            <Col className='col-2'>
                <button className='btn bg-primary text-light btn-block'>Search</button>
            </Col>
        </Row>
    </form>
  )
}

export default SearchForm