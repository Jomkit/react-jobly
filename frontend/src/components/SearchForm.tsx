import React, { useState } from 'react';
import JoblyApi from '../api.ts';
import { Col, FormGroup, Input, Label, Row } from 'reactstrap'

interface formDataInterface {
    name: string;
    minEmployees: number;
    maxEmployees: number;
}

const SearchForm = ({ setCompaniesData }: { setCompaniesData: Function }) => {
    const [ formData, setFormData ] = useState({
        name: "", 
        minEmployees: 0,
        maxEmployees: 10000
    });

    // Need to apply filter to form data
    async function getFilterCompanies(filterParams: formDataInterface) {
        try{
            const res = await JoblyApi.getCompanies(filterParams);
            setCompaniesData(res);
        } catch(e: any){
            console.log("Oops, something went wrong...");
            console.error(e);
        }
    }
    
    function handleChange(evt: React.ChangeEvent<HTMLInputElement>){
        const { name, value } = evt.target;
        setFormData((data: formDataInterface) => ({
            ...data,
            [name]: value,
        }));
    }

    function handleSubmit(evt: React.FormEvent){
        evt.preventDefault();
        console.log("Submitting...");
        getFilterCompanies(formData);
    }

  return (
    <form onSubmit={handleSubmit}>
        <Row className='m-lg-4 align-items-center'>
            <Col lg={6}>
                <FormGroup>
                    <Label
                        className='visually-hidden'
                        for='name'
                    >
                        Name
                    </Label>
                    <Input
                    id='name'
                    name='name'
                    placeholder='search companies'
                    type="text"
                    value={formData.name}
                    onChange={handleChange}
                    />            
                </FormGroup>
            </Col>
            <Col lg={2}>
                <FormGroup floating>
                    <Input 
                        id="minEmployees"
                        name="minEmployees"
                        placeholder='0'
                        min={0}
                        type="number"
                        bsSize='sm'
                        value={formData.minEmployees}
                        onChange={handleChange}
                    />
                    <Label for="minEmployees">Min Employees</Label>
                </FormGroup>
            </Col>
            <Col lg={2}>
                <FormGroup floating>
                    <Input 
                        id="maxEmployees"
                        name="maxEmployees"
                        placeholder='0'
                        min={0}
                        type="number"
                        bsSize='sm'
                        value={formData.maxEmployees}
                        onChange={handleChange}
                    />
                    <Label for="maxEmployees">Max Employees</Label>
                </FormGroup>
            </Col>

            <Col className='col-2'>
                <button className='btn bg-primary text-light btn-block'>Search</button>
            </Col>
        </Row>
    </form>
  )
}

export default SearchForm