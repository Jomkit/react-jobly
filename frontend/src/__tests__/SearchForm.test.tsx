import { fireEvent, render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import SearchForm from "../components/SearchForm";
import { FAKEDATA } from "./_testCommon";
import JoblyApi from "../api.ts";

vi.mock("../api.ts");
JoblyApi.getCompanies = vi.fn().mockResolvedValue(FAKEDATA);

const mockSetData = vi.fn();

it('should render without crashing', () => { 
    render(
        <MemoryRouter>
            <SearchForm setCompaniesData={mockSetData} />
        </MemoryRouter>
    )
})

it('should match snapshot', () => { 
    const { asFragment } = render(
        <MemoryRouter>
            <SearchForm setCompaniesData={mockSetData} />
        </MemoryRouter>
    )

    expect(asFragment()).toMatchSnapshot();
})

it('should call JoblyApi.getCompanies when search submit button pressed', () => { 
    const { getByText, getByLabelText } = render(
        <MemoryRouter>
            <SearchForm setCompaniesData={mockSetData} />
        </MemoryRouter>
    )

    const submitBtn = getByText("Search");
    const nameInput = getByLabelText("Name");
    fireEvent.change(nameInput, { target: { value: "test company 1" } });
    fireEvent.click(submitBtn);
    expect(JoblyApi.getCompanies).toHaveBeenCalled();
})