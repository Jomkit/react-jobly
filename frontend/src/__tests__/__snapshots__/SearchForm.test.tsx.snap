// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should match snapshot 1`] = `
<DocumentFragment>
  <form>
    <div
      class="m-lg-4 align-items-center row"
    >
      <div
        class="col-lg-6"
      >
        <div
          class="mb-3"
        >
          <label
            class="visually-hidden form-label"
            for="name"
          >
            Name
          </label>
          <input
            class="form-control"
            id="name"
            name="name"
            placeholder="search companies"
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        class="col-lg-2"
      >
        <div
          class="mb-3 form-floating"
        >
          <input
            class="form-control-sm form-control"
            id="minEmployees"
            min="0"
            name="minEmployees"
            placeholder="0"
            type="number"
            value="0"
          />
          <label
            class="form-label"
            for="minEmployees"
          >
            Min Employees
          </label>
        </div>
      </div>
      <div
        class="col-lg-2"
      >
        <div
          class="mb-3 form-floating"
        >
          <input
            class="form-control-sm form-control"
            id="maxEmployees"
            min="0"
            name="maxEmployees"
            placeholder="0"
            type="number"
            value="10000"
          />
          <label
            class="form-label"
            for="maxEmployees"
          >
            Max Employees
          </label>
        </div>
      </div>
      <div
        class="col-2 col"
      >
        <button
          class="btn bg-primary text-light btn-block"
        >
          Search
        </button>
      </div>
    </div>
  </form>
</DocumentFragment>
`;
